# PWSkills Assignment
Course: Decode Data Science With Machine Learning

Note:
1. Each Question Part consists of 3 Cells - (2 Markdown Cells, 1 each for question and the explanation or insights of the code & 1 Coding Cell with the Solution).
2. Each Code has been Executed and the Output can be seen just below the Coding Cell.

Question 1.3. Answer the following:

name = ["Mohan", "dash", "karam", "chandra","gandhi","Bapu"]

Do the following operations in this list:

a. Add an element "freedom_fighter" in this list at the 0th index.

b. Find the output of the following and explain how?

name = [“freedomFighter”,"Bapuji","MOhan" "dash", "karam", "chandra","gandhi"]
length1=len((name[-len(name)+1:-1:2]))
length2=len((name[-len(name)+1:-1]))
print(length1+length2)

c. Add two more elements in the name ["NetaJi","Bose"] at the end of the list.

d. what will be the value of temp?

name = ["Bapuji", "dash", "karam", "chandra","gandi","Mohan"]
temp=name[-1]
name[-1]=name[0]
name[0]=temp
print(name)

In [14]:
# The given List is:-

name = ["Mohan", "dash", "karam", "chandra", "gandhi", "Bapu"]

# Answer of a:
# To add "freedom_fighter" at the 0th index

name. insert (0, "freedom_fighter")

# Print the Updated List
print (f"The Updated List is:\n", name)

# Print a New Line
print ("\n")

# Answer of b:
# Find the Output of the following code

name = ["freedomFighter", "Bapuji", "MOhan", "dash", "karam", "chandra", "gandhi"]
length1 = len(name[-len(name)+1:-1:2])
length2 = len(name[-len(name)+1:-1])
print(length1 + length2)

# Print a New Line
print ("\n")

# Answer of c:
# Add two more elements in the list ["Netaji", "Bose"] at the end of the list

# The List we have
name = ["freedomFighter", "Bapuji", "MOhan", "dash", "karam", "chandra", "gandhi"]

# Adding elements at the end
name. extend (["Netaji", "Bose"])

# The New List
print (f"The Updated List is:\n", name)

# Print a New Line
print ("\n")

# Answer of d:
# Find the Value of 'temp'

# Given Code:
name = ["Bapuji", "dash", "karam", "chandra", "gandhi", "Mohan"]
temp = name[-1]
name[-1] = name[0]
name[0] = temp
print(name)

The Updated List is:
 ['freedom_fighter', 'Mohan', 'dash', 'karam', 'chandra', 'gandhi', 'Bapu']


8


The Updated List is:
 ['freedomFighter', 'Bapuji', 'MOhan', 'dash', 'karam', 'chandra', 'gandhi', 'Netaji', 'Bose']


['Mohan', 'dash', 'karam', 'chandra', 'gandhi', 'Bapuji']


Answer 1.3:

a. The Updated List is:
    ['freedom_fighter', 'Mohan', 'dash', 'karam', 'chandra', 'gandhi', 'Bapu']

b. Output: 8

Step by Step Explanation:-

    Step 1: List Initialization
        name = ["freedomFighter", "Bapuji", "MOhan", "dash", "karam", "chandra", "gandhi"]

        This list 'name' contains 7 elements.

    Step 2: Calculate 'length1'
        length1 = len(name[-len(name)+1:-1:2])

        '-len(name)' is -7, so 'len(name)+1' is -6; therefore, the slice 'name[-6:-1:2] implies a start from elements at index '-6' upto but not  including the element at '-1' index, with a step of 2.

        The breakdown of the above stated statement is given below:
        
            i. The starting index -6 corresponds to the element "Bapuji";
            ii. The ending index -1 corresponds to the element "gandhi", but it is not included in the slice;
            iii. With a step of 2, the slice name[-6:-1:2] includes - 
                "Bapuji" (index -6)
                "dash" (index -4)
                "chandra" (index -2)

        So, name[-6:-1:2] is:
            ["Bapuji", "dash", "chandra"]
        Therefore, length1 = len(["Bapuji", "dash", "chandra"]) = 3

    Step 3: Calculate 'length2'
        length2 = len(name[-len(name)+1:-1])

        As already stated above, '-len(name)' is -7, so '-len(name)+1' is -6, therefore, the slice 'name[-6:-1:2] implies a start from elements at index '-6' upto but not including the element at '-1'.

        The breakdown of the above stated statement is given below:
        
            i. The starting index -6 corresponds to the element "Bapuji";
            ii. The ending index -1 corresponds to the element "gandhi", but it is not included in the slice;
            iii. The slice name[-6:-1:2] includes - 
                "Bapuji" (index -6)
                "MOhan" (index -5)
                "dash" (index -4)
                "karam" (index -3)
                "chandra" (index -2)

        So, name[-6:-1] is:
            ["Bapuji", "MOhan", "dash", "karam", "chandra"]
        Therefore, length2 = len(["Bapuji", "MOhan", "dash", "karam", "chandra"]) = 5

    Step 4: Calculate the Sum
        print(length1 + length2)

        We have length1 = 3 and length2 = 5
    
        So, the sum is:
            3 + 5 = 8
    Therefore, the Output of the code is 8.

c. The Updated List is:
    ['freedomFighter', 'Bapuji', 'MOhan', 'dash', 'karam', 'chandra', 'gandhi', 'Netaji', 'Bose']

d. The List will be printed as follows:
    ['Mohan', 'dash', 'karam', 'chandra', 'gandhi', 'Bapuji']

    The Value of 'temp' will be "Mohan".

Question 1.4. Find the output of the following:

animal = ['Human','cat','mat','cat','rat','Human', 'Lion']
print(animal.count('Human'))
print(animal.index('rat'))
print(len(animal))

In [1]:
# Question 1.4:

# The given Code is as follows:

animal = ['Human','cat','mat','cat','rat','Human', 'Lion']
print(animal.count('Human'))
print(animal.index('rat'))
print(len(animal))

2
4
7


Answer 1.4.:
The output of the given code is:

Count (Frequency) of Human in the List animal = 2
Index (Position) of rat in the List animal = 4
Length (Number of elements) of the List animal = 7

Question 1.5.

tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no":1}, {"name":"Navneet"}])

Answer the following:

a. print(len(tuple1))

b. print(tuple1[-1][-1]["name"])

c. fetch the value of roll_no from this tuple.

d. print(tuple1[-3][1])

e. fetch the element "22" from this tuple.

In [15]:
# Qustion 1.5:

# The given Tuple is:

tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no":1}, {"name":"Navneet"}])

# Answer of a:
# Print the Length of the given Tuple 'tuple1'

print (len (tuple1))

# Print a New Line
print ("\n")

# Answer of b:
# Print the Last Element of the Tuple, which is associated with a key "name"

print (tuple1 [-1][-1]["name"])

# Print a New Line
print ("\n")

# Answer of c:
# Fetch the value of 'roll_no' from the given Tuple

ans_c = tuple1 [7][0]["roll_no"]
print (ans_c)

# Print a New Line
print ("\n")

# Answer of d:
# Fetch the Third Last Element of the Tuple and print its Second Element

print(tuple1[-3][1])

# Print a New Line
print ("\n")

# Ansswer of e:
# Fetch the element "22" from the given Tuple

ans_e = tuple1[6][2]
print(ans_e)

8


Navneet


1


ji


22


Answer 1.5:

a. The first question asks about the Length of the given Tuple.
    The Length of the Tuple 'tuple1' is 8.

b. The second question asks to access the last element of a tuple, and then access a dictionary within that element to retrieve the value associated with the key "name".

    The breakdown of the above statement is given as follows:
        i. tuple1[-1]: This accesses the last element of the tuple tuple1.
        ii. tuple1[-1][-1]: This accesses the last element of the inner tuple/list.
        iii. tuple1[-1][-1]["name"]: This accesses the dictionary within the inner list, present in the tuple and retrieves the value associated with the key "name".

    Therefore, the Output is "Navneet"

c. The third question asks to fetch the value of 'roll_no' from the given Tuple.

    The process can be broken down as follows:
        i. tuple1[7] gives the list [{"roll_no": 1}, {"name": "Navneet"}].
        ii. tuple1[7][0] gives the dictionary {"roll_no": 1}.
        iii. tuple1[7][0]["roll_no"] gives the value 1.
    
    Hence, the Output is 1 and the Index of the value 'roll_no' can be referred as tuple1 [7][0].

d. The fourth question asks to fetch the Third Last Element of the Tuple and print its Second Element. 

    The process can be broken down as follows:
        i. The tuple 'tuple1' contains several elements, including a list at index -3 (i.e., the third element from the end). This list is ["master", "ji"].
        ii. When 'tuple1 [-3][1]' is accessed, firstly, the element at index -3 of 'tuple1' is fetched, which is the list ["master", "ji"] and secondly, the element at index 1 of this list is fetched, which is "ji".
    
    So, the Output is "ji".

e. The fifth and the last question asks to fetch the value of '22' from the given Tuple.

    The process can be broken down as follows:
        i. tuple1[6] gives the tuple ("sita", "geeta", 22).
        ii. tuple1[6][2] gives the value 22.
    
    So, the Index of the value '22' can be referred as tuple1 [6][2].

Question 1.6:

Write a program to display the appropriate message as per the color of signal (RED-Stop/Yellow-Stay/ Green-Go) at the road crossing.

In [20]:
# Question 1.6:

"""
    Program to Display Appropriate Message according to the Color of the Signal at the Road Crossing.
    The Signal - Message Pair is given below:
        Red - Stop
        Yellow - Stay
        Green - Go
"""

def traffic_signal (color):
    if color.casefold () == "red":
        return "Stop"
    elif color.casefold () == "yellow":
        return "Stay"
    elif color.casefold () == "green":
        return "Go"
    else:
        return "Invalid Color.\nEnter a Valid Color!!!"

signal_color = input("Enter the color of the traffic signal: ")
message = traffic_signal (signal_color)
print (f"The Color of the Signal is: ", signal_color)
print(f"Message: ", message)

The Color of the Signal is:  Green
Message:  Go


Answer 1.6:

The objective of the above code is to Display Appropriate Message according to the Color of the Signal.

The explanation for the same is given below:
    i. The function traffic_signal (color) takes a color as input.

    ii. By using casefold() on the color variable, we ensure that the comparison is case-insensitive, so the program will work correctly regardless of whether the user enters the color in uppercase, lowercase, or a mix of both.
    
    iii. It checks the color of the signal and returns an appropriate message -
        If the color is "Red", it returns "Stop";
        if the color is "Yellow", it returns "Stay";
        If the color is "Green", it returns "Go";
        For any other Input, it returns "Invalid Color.\nEnter a Valid Color!!!"
    
    iv. The code prompts the user to input the color of the traffic signal and then prints both the Color of the Signal and the corresponding message.
    
    v. The Output of the code, for different scenarios are as follows -
        signal_color = "Red" : Output = "The Color of the Signal is: Red\nMessage: Stop"
        signal_color = "Yellow" : Output = "The color of the Signal is: Yellow\nMessage: Stay"
        signal_color = "Green" : Output = "The color of the Signal is: Green\nMessage: Go"
        For any other Input value of signal_color : Output = "Invalid Color.\nEnter a Valid Color!!!"


Question 1.7:

Write a program to create a simple calculator performing only four basic operations (+,-,/,*).

In [25]:
# Question 1.7:

# Program to create a Simple Calculator performing only four basic operations

# Functions for Operations

# Addition function
def sum (n1, n2):
    return n1 + n2

# Subtraction function
def subtract (n1, n2):
    return n1 - n2

# Multiplication function
def prod (n1, n2):
    return n1 * n2

# Division function
def div (n1, n2):
    
    # Check for the value of the Denominator
    if (n2 == 0):
        return "ERROR!!!\nDivision by 0 is NOT ALLOWED!!!\nPlease Enter a Non - Zero Number.\n"
    else:
        return n1 / n2

# Calculator function
def calculator ():
    print ("Welcome User!\nThis is a Simple Calculator.\n")
    print ("Select Operation:")
    print ("1. Addition (+)")
    print ("2. Subtraction (-)")
    print ("3. Multiplication (*)")
    print ("4. Division (/)")

    # Take User Input for the Type of Operation to be Performed
    choice = int (input ("Enter your Operation Choice: "))
    if choice < 1 or choice > 4:
        return "Invalid Choice.\nPlease Enter a Valid Choice.\n"
    
    # Exception Handling, in case of Invalid Input
    while True:
        try:
            num1 = float (input ("Enter the 1st Number: "))
            break
        except ValueError:
            print ("Invalid Input.\nPlease Enter a Valid Number.\n")
    
    while True:
        try:
            num2 = float (input ("Enter the 2nd Number: "))
            break
        except ValueError:
            print ("Invalid Input.\nPlease Enter a Valid Number.\n")
    
    operation = ""
    
    # Printing the USer Input
    print (f"\nThe 1st Number is: {num1}")
    print (f"The 2nd Number is: {num2}\n")
    
    # Printing the Output
    if choice == 1:
        operation = "Addition"
        print (f"Result of {operation}:\n{num1} + {num2} = {sum (num1, num2)}")
    elif choice == 2:
        operation = "Subtraction"
        print (f"Result of {operation}:\n{num1} - {num2} = {subtract (num1, num2)}")
    elif choice == 3:
        operation = "Multiplication"
        print (f"Result of {operation}:\n{num1} * {num2} = {prod (num1, num2)}")
    elif choice == 4:
        operation = "Division"
        print (f"Result of {operation}:\n{num1} / {num2} = {div (num1, num2)}")

# Run the Calculator
calculator ()

Welcome User!
This is a Simple Calculator.

Select Operation:
1. Addition (+)
2. Subtraction (-)
3. Multiplication (*)
4. Division (/)

The 1st Number is: 20.0
The 2nd Number is: 10.0

Result of Addition:
20.0 + 10.0 = 30.0


Answer 1.7:

The objective of the code is to create a Simple Calculator, capable of performing only basic 4 Arithmatic Functions, namely -
    a. Addition
    b. Subtraction
    c. Multiplication
    d. Division

Here's how the code works:
    
    i. Functions for Operation:
        a. 'sum (n1, n2)' : Returns the sum of 'n1' and 'n2'.
        b. 'subtract (n1, n2) : Returns the difference between 'n1' and 'n2'.
        c. 'prod (n1, n2) : Returns the product of 'n1' and 'n2'.
        d. 'div (n1, n2) : Returns the quotient of 'n1' divided by 'n2' and also an Error Message is displayed, if the Denominator is 0.
    
    ii. Calculator Function:
        a. Displays a menu with all the 4 available operations - Addition, Subtraction, Multiplication and Division.
        b. A relevant message for the same is also displayed.
        c. The user is prompted to enter their choice by entering a number (1, 2, 3, or 4) and an Error Message is also displayed, if the Input is not within the appropriate range.
        d. The user is then prompted to enter two numbers.
        e. Exception Handling is done and the user is prompted to re-enter a number if they enter a non-numeric value (e.g., a string or a special character).
        f. A variable named 'operation' is assigned to display the choice made by the user.
        g. The User Input is also printed.
        h. Performs the chosen operation and prints the result with an appropriate message.
    
    iii. Sample Outputs for the generated Code:-
        Let us take -
            num1 = 20.0
            num2 = 10.0
        a. Addition:
            Result of Addition:
            20.0 + 10.0 = 30.0
        b. Subtraction
            Result of Subtraction:
            20.0 - 10.0 = 10.0
        c. Multiplication:
            Result of Multiplication:
            20.0 * 10.0 = 200.0
        d. Division:
            Result of Division:
            20.0 / 10.0 = 2.0

Question 1.8.

Write a program to find the larger of the three pre-specified numbers using ternary operators.

In [27]:
# Question 1.8:

"""
    Program to find the larger of the three numbers
    Numbers are pre - specified
    Ternary Operators
"""

# Pre - Specified Numbers
n1 = 7
n2 = 18
n3 = 45

# Performing the Required Operation
result = n1 if (n1 > n2 and n1 > n3) else (n2 if n2 > n3 else n3)

# Display the Result
print (f"The Largest among {n1}, {n2} and {n3} is {result}")

The Largest among 7, 18 and 45 is 45


Answer 1.8:

Output: The Largest among 7, 18 and 45 is 45 [The numbers are subject to change, according to the preference of the user.]

Explanation:

i. Pre-specified numbers: We define three numbers a, b, and c.

ii. Ternary operation:
    a. The first ternary checks if a is greater than both b and c. If true, a is the largest.
    b. If the first condition is false, the second ternary checks if b is greater than c. If true, b is the largest. Otherwise, c is the largest.

iii. Display the result: The program prints the largest number.

Question 1.9.

Write a program to find the factors of a whole number  using a while loop.

In [28]:
# Question 1.9:

# Program to find the Factors of a Whole Number using Whole Number

# Function to find factors of a number using a while loop
def find_factors(number):
    factors = []
    i = 1
    while i <= number:
        if number % i == 0:
            factors. append (i)
        i += 1
    return factors

# Take User Input
number = int (input ("Enter a whole number: "))

# Print the Output, based on Condition
if number > 0:
    factors = find_factors (number)
    print(f"The factors of {number} are: {factors}")
else:
    print("Please enter a positive whole number.")

The factors of 45 are: [1, 3, 5, 9, 15, 45]


Answer 1.9:

Explanation:

i. Function to find Factors:
    a. 'find_factors(number)' : This function takes a whole number as input and returns a list of its factors.
    b. It initializes an empty list factors to store the factors.
    c. It uses a while loop to iterate from 1 to number (inclusive).
    d. Inside the loop, it checks if number is divisible by i. If true, it appends i to the list of factors.
    e. It increments i by 1 in each iteration.

ii. Take User Input: The program prompts the user to input a whole number.

iii. Check if the Number is Positive:
    a. The program checks if the entered number is positive. If it is, it calls the find_factors function and prints the factors.
    b. If the number is not positive, it prints a message asking for a positive whole number.

iv. Output:
    a. This program will output the factors of the entered whole number.
    b. Sample Output : The factors of 45 are: [1, 3, 5, 9, 15, 45] (The output is subject to change, depending upon the number entered by the user.)

Question 1.10.

Write a program to find the sum of all the positive numbers entered by the user. As soon as the user enters a negative number, stop taking in any further input from the user and display the sum.

In [29]:
# Question 1.10:

"""
    Program to find the Sum of all Positive Numbers
    Take Input from User
    Negative Number to STOP the Program
"""

# Function to Add the Positive Numbers
def sum_positive_numbers():
    total_sum = 0
    num = []
    
    # Exception Handling, in case of any Invalid Input
    while True:
        try:
            n = float(input("Enter a Number: "))
            num. append (n)
            if n < 0:
                break
            total_sum += n
        except ValueError:
            print("Please enter a valid number.")
    
    # Print the Output
    print (f"The Numbers Entered by the User are:\n{num}\n")
    print (f"The sum of all positive numbers entered is: {total_sum}")

# Call the function to start the program
sum_positive_numbers ()

The Numbers Entered by the User are:
[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, -69.0]

The sum of all positive numbers entered is: 55.0


Answer 1.10:

Explanation:

i. The function 'sum_positive_numbers()' initializes total_sum to 0.

ii. It enters an infinite loop (while True).

iii. Inside the loop, it prompts the user to enter a number.

iv. The try-except block is used to handle invalid inputs (e.g., non-numeric values).

v. If the entered number is negative, the loop breaks.

vi. If the entered number is positive, it adds the number to total_sum.

vii. All the values entered by the user are appended to a list.

viii. Output:
    a. When a negative number is entered, the loop stops, and the program prints the total sum of all positive numbers entered, along with the values entered by the user (including the negative number), in a list.
    b. Sample Output -
        The Numbers Entered by the User are:
        [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, -69.0]
        The sum of all positive numbers entered is: 55.0
        Note: The Output can change, depending upon the USer Input.

Question 1.11.

Write a program to find prime numbers between 2 to 100 using nested for loops.

In [38]:
# Question 1.11:

# Program to find Prime Numbers between 2 to 100, using nested for loops

# Import Module
import math

# Initialize an empty list to store prime numbers
prime_num_list = []

# Iterate over numbers from 2 to 100
for n in range(2, 101):
    is_prime = True

    # Check if the number is divisible by any number up to its square root
    for i in range(2, int (math. sqrt (n)) + 1):
        if n % i == 0:
            is_prime = False
            break
    if is_prime:
        prime_num_list.append(n)

# Print the Output
print(f"Prime numbers between 2 and 100:\n{prime_num_list}\n")
print (f"There are {(len (prime_num_list))} Prime Numbers within the Specified Range of 2 and 100.")

Prime numbers between 2 and 100:
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]

There are 25 Prime Numbers within the Specified Range of 2 and 100.


Answer 1.11:

The above code is designed to find and print all prime numbers between 2 and 100.

Here's a breakdown of how it works:
    
    i. Initialization: An empty list 'prime_num_list' is created to store the prime numbers found.
    
    ii. Main Loop: The code iterates over numbers from 2 to 100 using a for loop with range(2, 101).
    
    iii. Checking for Primality:
        a. For each number 'n' in the range, a flag 'is_prime' is set to "True", assuming it's prime initially.
        b. The code then checks if 'n' is divisible by any number 'i' between 2 and the square root of 'n' (inclusive) using another for loop. For finding the square root of 'n', math module has been used.
        c. If 'n' is divisible by any 'i' (i.e., 'n % i == 0'), the 'is_prime' flag is set to "False" and the inner loop is exited using the 'break' statement.
        d. If the 'is_prime' flag remains "True" after the inner loop, it means 'n' is a prime number, and it's appended to the "prime_num_list".
        e. The code repeats this process for all numbers between 2 and 100, and the resulting list of prime numbers is printed at the end.
    
    iv. Output:
        a. Finally, the code prints the list of prime numbers found, along with the count of prime numbers in the specified range.
        b. Sample Output -
            Prime numbers between 2 and 100:
            [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]
            There are 25 Prime Numbers within the Specified Range of 2 and 100.


Question 1.12

Write the  programs for the following:

a. Accept the marks of the student in five major subjects and display the same.

b. Calculate the sum of the marks of all subjects.Divide the total marks by number of subjects (i.e. 5), calculate percentage = total marks/5 and display the percentage.

c. Find the grade of the student as per the following criteria.
Hint: Use Match & case for this.

Criteria                          Grade
Percentage > 85                     A
Percentage < 85 and >= 75           B
Percentage < 75 and >= 50           C
Percentage > 30 and <= 50           D
Percentage < 30                  Reappear

In [2]:
# Question 1.12:

"""
    Program to take user input for marks in 5 subjects
    Calculate Total Marks and Percentage
    Find the Grade
"""
# Function to accept subjects and display them
def accept_subjects ():
    subjects = []
    for i in range (5):
        subject = input (f"Enter subject {i+1} name: ")
        subjects. append (subject)
    return subjects

# Function to accept marks of 5 subjects and display them
def accept_marks (subjects):
    marks = {}
    for subject in subjects:
        mark = float (input (f"Enter marks for {subject}: "))
        marks [subject] = mark
    return marks

# Function to calculate total marks
def calculate_total_marks (marks):
    total_marks = sum (marks. values ())
    return total_marks

# Function to calculate percentage
def calculate_percentage (total_marks):
    percentage = total_marks / 5
    return percentage

# Function to determine the grade based on percentage
def determine_grade (percentage):
    match percentage:
        case p if p > 85:
            grade = 'A'
        case p if p >= 75:
            grade = 'B'
        case p if p >= 50:
            grade = 'C'
        case p if p > 30:
            grade = 'D'
        case _:
            grade = 'Reappear'
    return grade

# Main function to run the program
def main ():
    subjects = accept_subjects ()
    print(f"Subjects: {subjects}\n")
    
    marks = accept_marks (subjects)
    print("Marks entered:")
    for subject, mark in marks. items ():
        print(f"{subject}: {mark}")
    
    total_marks = calculate_total_marks (marks)
    print (f"\nTotal Marks: {total_marks}")
    
    percentage = calculate_percentage (total_marks)
    print(f"\nPercentage: {percentage}%")
    
    grade = determine_grade (percentage)
    print (f"\nGrade: {grade}")

if __name__ == "__main__":
    main ()

Subjects: ['Mathematics', 'Physics', 'Chemistry', 'Biology', 'Computer Science']

Marks entered:
Mathematics: 90.0
Physics: 92.0
Chemistry: 91.0
Biology: 94.0
Computer Science: 95.0

Total Marks: 462.0

Percentage: 92.4%

Grade: A


Answer 1.12:

Explanation:-

i. Objective: This program is designed to take user input for marks in 5 subjects, calculate the total marks and percentage, and then determine the grade based on the percentage.

ii. Functions used in the Program:
    a. 'accept_subjects()' function: This function takes user input for the names of 5 subjects and returns a list of these subjects. It uses a for loop to iterate 5 times, asking the user to enter the name of each subject.
    b. 'accept_marks(subjects)' function: This function takes user input for the marks of each subject and returns a dictionary where the keys are the subject names and the values are the corresponding marks. It uses a for loop to iterate over the list of subjects, asking the user to enter the marks for each subject.
    c. 'calculate_total_marks(marks)' function: This function calculates the total marks by summing up all the values in the marks dictionary.
    d. 'calculate_percentage(total_marks)' function: This function calculates the percentage by dividing the total marks by 5.
    e. 'determine_grade(percentage)' function: This function determines the grade based on the percentage. The grades are:
        A: above 85%
        B: 75% to 85%
        C: 50% to 75%
        D: above 30% to 50%
        Reappear: below 30%
    f. 'main()' function: This is the main function that runs the program. It calls the above functions in sequence, prints the results, and finally determines the grade.

iii. Steps of the Program:
    a. The program asks the user to enter the names of 5 subjects.
    b. It then asks the user to enter the marks for each subject.
    c. The total marks are calculated by summing up all the marks.
    d. The percentage is calculated by dividing the total marks by 5.
    e. The grade is determined based on the percentage.
    f. The program prints the subjects, marks, total marks, percentage, and grade.
    g. The program uses Python's built-in input() function to get user input, and the print() function to display the results. It also uses a dictionary to store the subjects and marks, and the sum() function to calculate the total marks.

iv. Output: Sample Output (for this code) -
        Subjects: ['Mathematics', 'Physics', 'Chemistry', 'Biology', 'Computer Science']
        Marks entered:
        Mathematics: 90.0
        Physics: 92.0
        Chemistry: 91.0
        Biology: 94.0
        Computer Science: 95.0
        Total Marks: 462.0
        Percentage: 92.4%
        Grade: A
    Note: This is only a Sample Output and is subjected to change, based on the Inputs from the User.


Question 1.13:

Write a program for VIBGYOR Spectrum based on their Wavelength using Wavelength Range:

Color                       Wavelength (nm)
Violet                      400.0 - 440.0
Indigo                      440.0 - 460.0
Blue                        460.0 - 500.0
Green                       500.0 - 570.0
Yellow                      570.0 - 590.0
Orange                      590.0 - 620.0
Red                         620.0 - 720.0

In [8]:
# Question 1.13:

# Program to Determine the Color based on their Wavelength (VIBGYOR Spectrum)

# Function to Determine the Color
def find_color (wavelength):
    if 400.0 <= wavelength < 440.0:
        return "Violet"
    elif 440.0 <= wavelength < 460.0:
        return "Indigo"
    elif 460.0 <= wavelength < 500.0:
        return "Blue"
    elif 500.0 <= wavelength < 570.0:
        return "Green"
    elif 570.0 <= wavelength < 590.0:
        return "Yellow"
    elif 590 <= wavelength < 620.0:
        return "Orange"
    elif 620.0 <= wavelength <= 720.0:
        return "Red"
    else:
        return "Wavelength out of Visible Range!!!\nPlease Enter a Valid Wavelenth."

# Main Function
def main ():
    
    # Exception Handling
    try:
        wavelength = float (input ("Enter the Wavelength (in nm): "))
        color = find_color (wavelength)
        print (f"\nThe Color corresponding to {wavelength} nm is {color}")
    except ValueError:
        print ("Please Enter a Valid Numeric Value for Wavelength!!!")

if __name__ == "__main__":
    main ()


The Color corresponding to 570.0 nm is Yellow


Answer 1.13:

Code Explanation:-

i. Objective: This program is designed to take user input for wavelength of light in nm (nanometer), and then determine the color name corresponding to this wavelength.

ii. 'find_color (wavelength)' function:
    a. This function takes a wavelength as input and determines the corresponding color based on the given ranges.
    b. It uses a series of if-elif statements to check which range the wavelength falls into and returns the corresponding color.
    c. If the wavelength does not fall into any of the specified ranges, it returns "Wavelength out of visible range".

iii. Input the Wavelength:
    a. The program prompts the user to enter the wavelength in nanometers (nm).
    b. It tries to convert the input to a float. If the input is not a valid number, it catches the ValueError and prints an error message.

iv. 'if __name__ == "__main__"' function:
    a. This block ensures that the main function is only executed when the script is run directly (i.e., not when it's imported as a module by another script).
    b. In this block, the main function is called to execute the code. Therefore, whenthis script is executed, it will prompt to enter the wavelength and then print the corresponding color in the VINGYOR Spectrum.

v. Output:
    a. This program will output the color in the VIBGYOR spectrum based on the given wavelength.
    b. Sample Output - The Color corresponding to 570.0 nm is Yellow (This is subjected to change as per User Input).

Question 1.14:

Consider the gravitational interactions between the Earth, Moon, and Sun in our solar system.

Given:

mass_earth = 5.972e24  # Mass of Earth in kilograms
mass_moon = 7.34767309e22  # Mass of Moon in kilograms
mass_sun = 1.989e30  # Mass of Sun in kilograms
distance_earth_sun = 1.496e11  # Average distance between Earth and Sun in meters
distance_moon_earth = 3.844e8  # Average distance between Moon and Earth in meters

Tasks:

    a. Calculate the gravitational force between the Earth and the Sun
    b. Calculate the gravitational force between the Moon and the Earth
    c. Compare the calculated forces to determine which gravitational force is stronger
    d. Explain which celestial body (Earth or Moon) is more attracted to the other based on the comparison

In [12]:
# Question 1.14:

"""
    Program to calculate the Gravitational Force between Earth and Sun, as well as Moon and Earth
    Compare the calculated forces to determine which Gravitational Force is Stronger
    Explain which celestial body (Earth or Moon) is attracted to the other, based on the comparison
"""

# Import Module
import math

# Given Values
mass_earth = 5.972e24  # Mass of Earth in kilograms
mass_moon = 7.34767309e22  # Mass of Moon in kilograms
mass_sun = 1.989e30  # Mass of Sun in kilograms
distance_earth_sun = 1.496e11  # Average distance between Earth and Sun in meters
distance_moon_earth = 3.844e8  # Average distance between Moon and Earth in meters

# Constant Value, but not given
G = 6.674e-11  # Gravitational constant in N(m^2)/kg^2

# Calculations

# Gravitational force between the Earth and the Sun
force_earth_sun = G * (mass_earth * mass_sun) / (distance_earth_sun**2)

# Gravitational force between the Moon and the Earth
force_moon_earth = G * (mass_moon * mass_earth) / (distance_moon_earth**2)

# Results
print ("The gravitational force between the Earth and the Sun is {:.2e} N". format (force_earth_sun))
print ("The gravitational force between the Moon and the Earth is {:.2e} N". format (force_moon_earth))

# Conclusions

# Compare the calculated forces
if force_earth_sun > force_moon_earth:
    print ("The gravitational force between the Earth and the Sun is stronger.")
else:
    print ("The gravitational force between the Moon and the Earth is stronger.")

# Determine which celestial body is more attracted to the other
if force_earth_sun > force_moon_earth:
    print ("The Earth is more attracted to the Sun than the Moon is to the Earth.")
else:
    print ("The Moon is more attracted to the Earth than the Earth is to the Sun.")

The gravitational force between the Earth and the Sun is 3.54e+22 N
The gravitational force between the Moon and the Earth is 1.98e+20 N
The gravitational force between the Earth and the Sun is stronger.
The Earth is more attracted to the Sun than the Moon is to the Earth.


Answer 1.14:

Code Explanation:-

i. Objective:
    a. This code calculates the gravitational forces between the Earth and the Sun, and the Moon and the Earth, and then compares these forces to determine which is stronger.
    b. It also explains which celestial body is more attracted to the other based on the comparison.

ii. Initialization:
    a. First, the 'math' module is imported (the only necessary module)
    b. The given and constant values are defined. The given values are the masses of the Earth, Moon, and Sun, as well as the average distances between the Earth and Sun, and the Moon and Earth. The constant value is the gravitational constant, G.

iii. Calculation: The gravitational forces between the Earth and the Sun, and the Moon and the Earth, are calculated, using the formula:
                    "F = G * (m1 * m2) / r^2"
                    where,
                        F is the gravitational force,
                        G is the gravitational constant,
                        m1 and m2 are the masses of the two objects, and
                        r is the distance between them.

iv. Printing: The calculated forces are then printed, and the code compares the two forces to determine which is stronger. If the force between the Earth and the Sun is greater, the code prints that the Earth is more attracted to the Sun than the Moon is to the Earth. If the force between the Moon and the Earth is greater, the code prints that the Moon is more attracted to the Earth than the Earth is to the Sun.

v. Outcome:
    a. The gravitational force between Earth and Sun should be significantly stronger than the gravitational force between Moon and Earth due to the massive difference in mass between the Sun and the Moon, despite the larger distance between the Earth and Sun. This means that the Earth is more attracted to the Sun based on gravitational force strength.
    b. Sample Outcome -
        The gravitational force between the Earth and the Sun is 3.54e+22 N
        The gravitational force between the Moon and the Earth is 1.98e+20 N
        The gravitational force between the Earth and the Sun is stronger.
        The Earth is more attracted to the Sun than the Moon is to the Earth.

Question 1 has been solved!!!